### DISCUSSION TITLE
Wireshark TREL packets decryption - No encryption key set -

### DISCUSSION THREAD

**Discussion by Yunagiri** [openthread/openthread#9267] [Q&A] [answered]
**Describe the bug** A clear and concise description of what the bug is.

TREL key doesn't allow packet decryption

**To Reproduce** Information to reproduce the behavior, including:

Hi, I am making a wireshark dissector for TREL packets. The idea is to parse the TREL header and chain the 802.15.4 dissector on the remaining tvb. TREL uses a hashed then expanded version of the Thread network key, that I retrieve through 

`void KeyManager::ComputeTrelKey(uint32_t aKeySequence, Mac::Key &aKey)`

It works on commissioning packets (Parent Requests/Responses, Child ID Request/Response are visible) but when I try to send raw udp messages through ot-ctl I still get the error: 

Expert Info (Warning/Undecoded): No encryption key set - can't decrypt

So either the TREL key used by `ComputeTrelKey()` is wrong (I took care to get the real TREL key and not the temporary one) or the data uses another key yet again, different from management/commissioning frames.

I'd love some pointers, thanks in advance.

1. Git commit id : df5b578e890b56c67d5beed849943e63a6fea56d
2. IEEE 802.15.4 hardware platform - ot-efr32 / ot-cli-ftd
4. Network topology : Leader <--TREL--> Router

**Expected behavior** A clear and concise description of what you expected to happen.

I expected to be able to decrypt TREL packets

**Console/log output** If applicable, add console/log output to help explain your problem.

**Additional context** Add any other context about the problem here.

![image](https://github.com/openthread/openthread/assets/44071344/9cdddf90-ad54-4a72-b152-05a56f43c573)

![image](https://github.com/openthread/openthread/assets/44071344/8114e85c-c0fd-4cbf-b097-1a010390241a)

**Comment by jwhui** [comment]
`KeyManager::ComputeTrelKey()` should generate the AES-128 key used to secure TREL frames.

Note that frames carrying MLE and Thread Commissioning packets do not have security enabled.

I'm not an expert in Wireshark dissectors, but my guess is that the AES key retrieval in Wireshark is not happening as expected.

**Reply by Yunagiri** [reply]
Wireshark's 802.15.4 dissector recognizes the header and the protocol, and just do a pass of all the keys in its table which I can freely add/remove 

![image](https://github.com/openthread/openthread/assets/44071344/634817b9-2e2a-49f0-99fe-e6c4b130d91a)
![image](https://github.com/openthread/openthread/assets/44071344/f2a6c9b9-92fe-4387-992d-2e7c73f75cb8)

So I don't think there are any problems possible on that side as I've always been able to decrypt normal Thread packets. 

Key retrieval is done by directly plugging this code into `KeyManager::ComputeTrelKey(` :

```
    char aKeyHexStr[2 * Mac::Key::kSize + 1];
    for (int i = 0; i < Mac::Key::kSize; i++) {
        sprintf(&aKeyHexStr[2* i], "%02X", aKey.m8[i]);
    }
    aKeyHexStr[2 * Mac::Key::kSize] = '\0';
    otLogInfoPlat("TREL key is : %s", aKeyHexStr);
```

**Reply by jwhui** [reply, marked as answer]
As currently implemented in Wireshark ,the `Decryption key` value is the Thread Network Key and the `Key hash` value indicates using Thread's key derivation function for generating the MLE and MAC keys. Thread-over-Infrastructure defines its own key derivation function, which is implemented by `KeyManager::ComputeTrelKey()`, so that would need to be implemented in Wireshark as well.

**Reply by Yunagiri** [reply]
Thank you for the hint! I knew I was missing something about MAC key derivation from network key, `KeyManager::ComputeTrelKey()` gives the TREL MAC key and so I had it raw, no further hashing required. Selecting the no hash option in `Key hash` allowed Wireshark to decrypt packets. For now, I deem it "good enough" for my purposes. The "real" implementation into the Wireshark's built-in 802.15.4 dissector with a separate TREL option looks abit more daunting.

![image](https://github.com/openthread/openthread/assets/44071344/bc9e7867-cb73-4716-9b49-5cd923cb2dbd)

One last question: As the Thread network key is used for MAC/MLE key generation, why TREL key only generates MAC key ? And how come I am able to decrypt CoAP packets aswell without the MLE key?

**Reply by abtink** [reply]
> As the Thread network key is used for MAC/MLE key generation, why TREL key only generates MAC key ? 

TREL is basically another radio link, so we only need a link/MAC layer key for it.
MLE uses UDP and is above MAC and IPv6 in terms of network layers. It uses its own security mechansim (MLE messages do not always use link level security and use MLE key to secure message at MLE layer). MLE message can be sent over any link (TREL or 15.4).

> And how come I am able to decrypt CoAP packets as well without the MLE key?

 MLE Key is only used for MLE message. CoAP uses link-level security.

**Reply by Yunagiri** [reply]
> TREL is basically another radio link, so we only need a link/MAC layer key for it. MLE uses UDP and is above MAC and IPv6 in terms of network layers. It uses its own security mechansim (MLE messages do not always use link level security and use MLE key to secure message at MLE layer). MLE message can be sent over any link (TREL or 15.4).

Hello and thank you for the precisions, if I understood correctly I can see most MLE packets because they are using the MAC/link key instead of the MLE key. What kind of MLE message actually uses the MLE key then? and ultimately do I need MAC key + MLE key to decrypt these MLE messages sent over TREL? Sorry last question just spawned more questions 😃 

> MLE Key is only used for MLE message. CoAP uses link-level security.

Right, I thought as CoAP was above IPv6 and UDP it would also have some kind of higher security for TMF frames. Many thanks for the precisions again!

**Reply by jwhui** [reply]
MAC and TREL are both link layer. MAC frames are secured using the MAC key, TREL frames are secured using the TREL key.

MLE key operates about the UDP layer and are always secured using the MLE key.

**Reply by abtink** [reply]
> I can see most MLE packets because they are using the MAC/link key instead of the MLE key. What kind of MLE message actually uses the MLE key then?

In general, MLE messages do not use MAC/link level security. Instead, they use MLE layer security, which uses the MLE key. The only case that MAC/link level security is used (in addition to MLE layer security) is when the message is too long and needs to be fragmented at 6LoWPAN layer and sent as multiple frames. 

Section 7.3 of the Thread specification describes MLE security in detail.

> ultimately do I need MAC key + MLE key to decrypt these MLE messages sent over TREL?

To decrypt TREL frames, we need the TREL key. Once the frame is processed through the 6LoWPAN layer and we get an IPv6 message, if it happens to be an MLE message (a UDP message using the MLE port number), then its payload can be decrypted using the MLE key. This will allow us to get the MLE message contents, such as the included MLE TLVs. I believe Wireshark can already detect MLE messages and perform the decryption.
